### PR TITLE
[MNT] CI steps to test by estimator with estimator specific dependencies, controlled by tag for separate VM

### PR DESCRIPTION
**Pull request by julian-fong**
related to #7087

### CODE REVIEW

**Comment by julian-fong** [comment]
@geetu040 any idea on how to fix the failures for `TimesFMForecaster`?

**Comment by geetu040** [comment]
> @geetu040 any idea on how to fix the failures for `TimesFMForecaster`?

the errors I see, seem to come from the files of chronos

**Comment by julian-fong** [comment]
@fkiraly Do you know why we're looking at the `chronos` forecaster when we're checking the estimator dependencies of `TimesFMForecaster` here https://github.com/sktime/sktime/actions/runs/15398624766/job/43328256474?pr=8331 ?

**Review by fkiraly — CHANGES_REQUESTED** [review]
From the test log, it seems that the environment is not set up properly?

**Comment by julian-fong** [comment]
> From the test log, it seems that the environment is not set up properly?

Ah yeah i was playing around with the ordering of the new code to see the interaction with the CI. I was trying to see if we place the new code before installing the dependencies for the given forecaster, could we skip the test before we begin installing the forecaster dependencies

**Comment by fkiraly** [comment]
excited about this PR - please let us know when it is ready to review!

**Comment by julian-fong** [comment]
Hi @fkiraly , I forgot to provide an update on the status for the pr - it is ready for a review!

**Comment by fkiraly** [comment]
Could you kindly explain what causes all the failures? Are they unrelated?

Plus, what is exactly the logic on spawning VM?
* which models get their own VM?
* when do these models get their own VM?
* how do we address the aim to not spawn too many VM in the same PR?

**Comment by julian-fong** [comment]
> Could you kindly explain what causes all the failures? Are they unrelated?
> 

The failures are unrelated - i believe we have seen these errors before in other prs

> Plus, what is exactly the logic on spawning VM?
> 
> * which models get their own VM?

The logic from the new tag will determine which models get their own VM via the code you have written inside `_get_all_changed_classes` and the changes to the CI (see below) which checks the tag `test_vm`. So if the estimator has the tag `test_vm` set to True, the model will get their own VM.

> * when do these models get their own VM?

The models get their own VM as long as the `test_vm` tag is set to True. Once this is set, we get the list of estimators that have their modules changed (via base class or the class itself) 

```python
OBJ_LIST=$(python -c 'from sktime.tests.test_switch import _get_all_changed_classes; import json; obj_list = _get_all_changed_classes(vm=True); print(json.dumps(obj_list))')
          OBJ_LIST_LENGTH=$(python -c 'from sktime.tests.test_switch import _get_all_changed_classes; obj_list = _get_all_changed_classes(vm=True); print(len(obj_list))')
echo "OBJ_LIST=$OBJ_LIST" >> $GITHUB_ENV
echo "OBJ_LIST_LENGTH=$OBJ_LIST_LENGTH" >> $GITHUB_ENV
echo "::set-output name=obj_list::$OBJ_LIST"
echo "::set-output name=obj_list_length::$OBJ_LIST_LENGTH"
```

and then we will check estimator dependencies (the additional code i've added in this pr) and run tests on their own VM as necessary.

> * how do we address the aim to not spawn too many VM in the same PR?

We can leverage the tag inside #8330 to do this via combination

If `test_vm` and `test_only_estimator` are both set to True, then the estimator will get its own VM if and only if the code in the estimator is changed, this will reduce the number of VMs that are being created in other prs that are not necessarily related to the estimator

if `test_vm` is set to True and `test_only_estimator` is set to False, then the estimator will be tested and get its own VM if and only if the estimator or classes it inherits are changed.

Obviously as well, reducing the number of estimators that have the `test_vm` tag set to True will reduce the number of VMs spawned.

**Comment by fkiraly** [comment]
More questions:

* the models that get their own VM are not tested anymore inside the standard CI?
* does this 

> then the estimator will get its own VM if and only if the code in the estimator is changed

mean that the estimator is no longer tested if the base class changes but not the estimator code, or a parent? (I would approve, although we should think of the edge case of intermediate base classes)

**Comment by julian-fong** [comment]
> * the models that get their own VM are not tested anymore inside the standard CI?

This should be the intended case, although not exactly sure how to verify this is the current behaviour

> * does this mean that the estimator is no longer tested if the base class changes but not the estimator code, or a parent? (I would approve, although we should think of the edge case of intermediate base classes)

Also yes, even if the base class changes or a parent class, we don't trigger tests. This tag should really be used as rarely as possible, and should only be tested on estimators that require large compute times in the CI

Regarding point number 2, do you have any scenarios that you have on top of your mind that may cause some concerns?

**Comment by fkiraly** [comment]
> This should be the intended case, although not exactly sure how to verify this is the current behaviour

I suppose, running the CI with a failure in an estimator? I would suggest branching off this PR and introducing a failure in the estimator, then see if it fails on the ordinary CI or not.

**Comment by fkiraly** [comment]
> Regarding point number 2, do you have any scenarios that you have on top of your mind that may cause some concerns?

Not in the current code base, it is just edge case thinking.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think the TimesFM is still failing, I think this is simply since the environment cannot be installed.

This is due to the linux platform requirement?

**Comment by julian-fong** [comment]
> I think the TimesFM is still failing, I think this is simply since the environment cannot be installed.
> 
> This is due to the linux platform requirement?

Yes I believe so. This is one of the drawbacks that I am seeing with the current code. I think it would be a good idea for us to check `env_marker` before installing soft dependencies, so we can skip the test on the VM if it doesn't satisfy the environment (just the os check condition). What do you think?

**Comment by fkiraly** [comment]
I think there are two routes:

* do not spin up the VM in the first place that are incompatible with markers and python bounds
* spin them up, but ensure than nothing gets installed or tested on them

I would go with whichever turns out to be easier, though I am not sure which that is.

**Comment by fkiraly** [comment]
FYI, I updated this with the changes from https://github.com/sktime/sktime/pull/7087

Is it correct that there was only a single incremental change - adding the `python>=3.11` condition to one of the foundation models?

**Comment by julian-fong** [comment]
> FYI, I updated this with the changes from #7087
> 
> Is it correct that there was only a single incremental change - adding the `python>=3.11` condition to one of the foundation models?

Yes this is correct, this was my only change so far

File: sktime/forecasting/chronos.py [hunk 1]
```diff
@@ -293,7 +293,7 @@ class ChronosForecaster(_BaseGlobalForecaster):
         ],
         # abdulfatir and lostella for amazon-science/chronos-forecasting
         "maintainers": ["geetu040"],
-        "python_dependencies": ["torch", "transformers", "accelerate"],
+        "python_dependencies": ["torch", "transformers", "accelerate", "numpy==1.26.4"],
```
Comment (fkiraly) [inline sktime/forecasting/chronos.py]: I assume this is just diagnostic? Because we cannot merge it like that, that would restrict `numpy` to a pin on all operating systems.

Comment (julian-fong) [reply sktime/forecasting/chronos.py]: yes!

**Comment by fkiraly** [comment]
this seems to work now! What was the trick?

File: sktime/forecasting/timesfm_forecaster.py [hunk 2]
```diff
@@ -180,7 +181,7 @@ class TimesFMForecaster(_BaseGlobalForecaster):
         "capability:pred_int:insample": False,
         "capability:global_forecasting": True,
         # testing configuration
-        # ---------------------
+        # ----------------------
```
Comment (fkiraly) [inline sktime/forecasting/timesfm_forecaster.py]: why are you changing this line?

File: sktime/forecasting/ttm.py [hunk 3]
```diff
@@ -216,6 +216,7 @@ class TinyTimeMixerForecaster(_BaseGlobalForecaster):
         # ajati, wgifford, vijaye12 for ibm-granite code
         "maintainers": ["geetu040"],
         "python_dependencies": ["transformers", "torch", "accelerate>=0.26.0"],
+        "python_version": ">=3.11",
```
Comment (fkiraly) [inline sktime/forecasting/ttm.py]: is this correct? Or does the original code run run on all python versions?

This is quite restrictive, so maybe you want to work on this in another PR?

File: sktime/forecasting/timesfm_forecaster.py (hunk 2 above)
Comment (julian-fong) [reply sktime/forecasting/timesfm_forecaster.py]: sorry, this line was added originally to trigger `timesfm` tests on the CI - i will remove it now!

**Comment by julian-fong** [comment]
> this seems to work now! What was the trick?

For `timesfm` it seems like the workers on the CI were installing older version of packages - which would break downstream code, so forcing the workers to install more recent package versions (for tensorflow, jax and jaxlib) seemed to fix this. Note that we only have one usable env for `timesfm` that being python 3.10 and ubuntu

For `ttm`. using the `macos-latest` worker does not have `mps` support, so you need to disable it somehow when running tests on the macos environments in the CI. I did this by turning off `mps` via a torch method in the `get_test_params` method. That way, users who are using the sktime `ttm` module on mac won't be affected, as it will only turn off `mps` when the tests are triggered via the CI. @fkiraly do you think this is a good workaround?

**Comment by julian-fong** [comment]
@fkiraly This is ready for review

**Comment by julian-fong** [comment]
> Great, seems to work!
> 
> Two things I want to explicitly check:
> 
> * have you checked that the new timesfm bounds are lenient enough that still the tests get executed on at least one VM in `test-est`? (this is important!)

Thanks for the heads up! I realized that changing the versions of the dependencies will not have any test executed for any environment in `TimesFM`, so I ended up reverting them. The only environment capable of running tests is `ubuntu-latest` with `python 3.10`, and I'm not able to resolve the genuine error.

> * have you checked whether the timesfm bounds on python or operating system could be relaxed? (maybe out of scope for this PR)

I doubt it. I looked at their main github page and they have the same dependencies on packages and python version

Other than that, the new code seems to work..

**Comment by fkiraly** [comment]
I think the error is due to `_safe_import` not finding an imported object. Could you try backtrack which object is not getting properly imported?

**Comment by julian-fong** [comment]
> I think the error is due to `_safe_import` not finding an imported object. Could you try backtrack which object is not getting properly imported?

I'm referring to the error in the ubuntu/python 3.10 test here:

https://github.com/sktime/sktime/actions/runs/16530320481/job/46753907298?pr=8331

the other error I believe is due to the `MomentFMForecaster` not using `_safe_import` as you mentioned - I can fix it in another pr if you think its necessary here

**Comment by fkiraly** [comment]
> I'm referring to the error in the ubuntu/python 3.10 test here:

Yes, so am I - a mock object is involved, which is the result of a failed `_safe_import`

**Review by fkiraly — CHANGES_REQUESTED** [review]
Could you copy the changes to the CI in the `.github` folder to another PR to quickly merge and fix the issues with various foundation model CI?

We can then look at the detail change to foundation models separately.

**Comment by julian-fong** [comment]
> Could you copy the changes to the CI in the `.github` folder to another PR to quickly merge and fix the issues with various foundation model CI?
> 
> We can then look at the detail change to foundation models separately.

ok will do

**Comment by julian-fong** [comment]
> > I'm referring to the error in the ubuntu/python 3.10 test here:
> 
> Yes, so am I - a mock object is involved, which is the result of a failed `_safe_import`

Is it possible to get some guidance on how to figure out which failed import it is? Not too familiar with magic mock
